### ISSUE TITLE
Show undecryptable file names in the dir listing

### ISSUE THREAD

**Issue by rfjakob** [rfjakob/gocryptfs#393]
Dropbox etc handle synchronisation conflicts by preserving both versions and appending something like `(conflict)` to the file name. 

The encrypted file will look like `9sv5JFmMzhvBE7faKVZZdQ (conflict)`. This does not decrypt properly, and gocryptfs will log an error and ignore the file. So basically, those `(conflict)` files become invisible.

We should find a way to make these files visible to the user. Simplest option is listing them as something like
```
GOCRYPTFS_INVALID_NAME_1234 9sv5JFmMzhvBE7faKVZZdQ (conflict)
```

**Comment by rfjakob** [comment]
Looking at 
* https://help.dropbox.com/syncing-uploads/conflicted-copy
* https://docs.nextcloud.com/desktop/2.3/conflicts.html
both seem to append ` (*` to the name. It might make sense to try and decrypt everything in front of the ` (`, as suggested in https://github.com/rfjakob/gocryptfs/issues/345#issuecomment-465306785 . The problem is, however, that it gives an attacker a way to insert arbitrary file names into the plaintext directory listing.

**Comment by brainchild0** [comment]
Overall, there are three cases:
1. Encrypted names
2. Plaintext names.
3. Mixed names.

Currently, only (1) is supported.  For the suggested behavior to occur, (3) must also be supported.  Would (2) also be supported?  

What specifically are the attack scenarios opened by (2) or (3) that are not already possible under (1)?  Is (3) safer than (2), or are both vulnerable?  Does (3) really have any more vulnerability than (1)?

Note also, I would suggest processing the names in a way that is general enough to accommodate the widest possible variety of naming conventions, not simply whichever is used by Dropbox.

**Comment by chairraver** [comment]
Hi, I'm experiencing a similar issue like in the original post. I'm using OneDrive Linux client https://github.com/abraunegg/onedrive for the actual storage syncing,  which appends a host name to the encrypted file name of the conflicting file. So, it appears for these conflicting situation, some kind of manual resolution option would be useful.

Let me also express my gratefulness for this marvellous piece of software. Thanks.

**Comment by DerDonut** [comment]
+1 for this feature. It would be the last piece I would need for my "perfect" personal data storage solution. Although I am not using online services. I would like to use [Syncthing](https://syncthing.net/) in a local network and this feature would allow to handle the conflict files.

Maybe it is more safe to introduce a parameter which configures the possible appendix which is used to display undecryptable filenames at mount? That would drastically reduce the possibilities to generate different filename which can be injected? Something regex-like, i.e.
`gocryptfs /cipher /plain -o namedecr='\(.{7,9}\)'`
If the parameter is not used, this feature is turned off completely

I was going to set a Bounty on that on https://www.bountysource.com, but unfortunately, they only offer Paypal. Following [another post](https://github.com/rfjakob/gocryptfs/issues/55), there is no plan to accept any other donations. If that changes, please let me know. I really want to contribute to your great work.

**Comment by slackner** [comment]
What about providing such undecryptable files in a separate folder like `lost+found`?

It could either be done globally (i.e., encountering an undecryptable file tracks it, and adds/moves it to `lost+found`) or it could be a hidden subfolder `.lost+found` that appears whenever a folder contains corrupted filenames.

**Comment by DerDonut** [comment]
Great.
Just compiled gocryptfs. gocryptfs -version now gives: `gocryptfs v1.7.1-36-g2568518; go-fuse v2.0.2-55-gfe141f3; 2020-03-01 go1.14 linux/amd64`
I mounted a share with 
`gocryptfs -badname=* <CIPHER> <MOUNTPOINT>`
Besides the already existing files the cipherdir containts:

> Folder .stfolder
> File sq7nHcTyK94IMrRYnKPxxQ
> File sq7nHcTyK94IMrRYnKPxxQ(copy)

When I now look into the plain dir, ls -al gives me:
```
ls: cannot access 'GOCRYPTFS_BAD_NAME .stfolder': No such file or directory
total 44
drwxrws---  9 donut donut  4096 Mar  1 23:31  .
drwxrws---  9 donut donut  4096 Mar  1 23:31  ..
d?????????  ? ?     ?         ?            ? 'GOCRYPTFS_BAD_NAME .stfolder'
-rw-rw----  1 donut donut     2 Mar  1 23:20  test_file.txt
```
sq7nHcTyK94IMrRYnKPxxQ is "test_file.txt". So, the folder ".stfolder" is visible (in terminal, not in Dolphin) but seems to be inaccessible. The invalid file name is not visible at all. Is my usage of the parameter **badname** wrong?

**Comment by DerDonut** [comment]
I just discovered that you used filepath.Match and not regexp.Match. Using the [corresponding syntax](https://golang.org/pkg/path/filepath/#Match), the visibility seems to work.
But it is still unclear why the content cannot be accessed. I tested both plain and cipher files inserted into cipher dir, same result.

Another question: Does anything stand against using regexp concerning security or speed? If not, I would propose a version in which the expression is separated and the remaining name is decrypted.
